ulog.cpp

```diff
@@ -0,0 +1,350 @@
+            = (struct mavlink_router_mavlink2_header *)buffer->data;
+        msg_id = msg->msgid;
+        payload_len = msg->payload_len;
+    } else {
+        struct mavlink_router_mavlink1_header *msg
+            = (struct mavlink_router_mavlink1_header *)buffer->data;
+        msg_id = msg->msgid;
+        payload_len = msg->payload_len;
+    }
+
+    /* Check if we are interested in this msg_id */
+    if (msg_id != MAVLINK_MSG_ID_COMMAND_ACK && msg_id != MAVLINK_MSG_ID_LOGGING_DATA_ACKED
+        && msg_id != MAVLINK_MSG_ID_LOGGING_DATA) {
+        return buffer->len;
+    }
+
+    uint8_t *payload;
+
+    if (mavlink2) {
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink2_header);
+        trimmed_zeros = get_trimmed_zeros(buffer);
+    } else {
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink1_header);
+        trimmed_zeros = 0;
+    }
+
+    /* Handle messages */
+    switch (msg_id) {
+    case MAVLINK_MSG_ID_COMMAND_ACK: {
+        mavlink_command_ack_t cmd;
+
+        memcpy(&cmd, payload, payload_len);
+        if (trimmed_zeros)
+            memset(((uint8_t *)&cmd) + payload_len, 0, trimmed_zeros);
+
+        if (!_timeout || cmd.command != MAV_CMD_LOGGING_START)
+            break;
+
+        if (cmd.result == MAV_RESULT_ACCEPTED) {
+            _mainloop->del_timeout(_timeout);
+            _timeout = NULL;
+        } else
+            log_error("MAV_CMD_LOGGING_START result(%u) is different than accepted", cmd.result);
+        break;
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA_ACKED: {
+        mavlink_logging_data_acked_t *ulog_data_acked = (mavlink_logging_data_acked_t *)payload;
+        mavlink_message_t msg;
+        mavlink_logging_ack_t ack;
+
+        ack.sequence = ulog_data_acked->sequence;
+        ack.target_component = MAV_COMP_ID_ALL;
+        ack.target_system = _target_system_id;
+        mavlink_msg_logging_ack_encode(SYSTEM_ID, MAV_COMP_ID_ALL, &msg, &ack);
+        _send_msg(&msg, _target_system_id);
+        /* no break needed, message will be handled by MAVLINK_MSG_ID_LOGGING_DATA case */
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA: {
+        if (trimmed_zeros) {
+            mavlink_logging_data_t ulog_data;
+            memcpy(&ulog_data, payload, payload_len);
+            memset(((uint8_t *)&ulog_data) + payload_len, 0, trimmed_zeros);
+            _logging_data_process(&ulog_data);
+        } else {
+            mavlink_logging_data_t *ulog_data = (mavlink_logging_data_t *)payload;
+            _logging_data_process(ulog_data);
+        }
+        break;
+    }
+    }
+
+    return buffer->len;
+}
+
+bool ULog::_logging_seq(uint16_t seq, bool *drop)
+{
+    /* Check for message drops */
+    if (_expected_seq != seq) {
+        *drop = true;
+
+        /*
+         * Handle case were Flight Stack did not received a ACK and sent the
+         * same message again.
+         */
+        if (_expected_seq - 1 == seq)
```

Comment:
this doesn't seem to handle the case @bkueng commented here: https://github.com/01org/mavlink-router/pull/10/files#r100004968 - "A duplicate could be further back than just the previous message"